### DISCUSSION TITLE
DNS errors when accessing dynamically added app running in a kubernetes cluster

### DISCUSSION THREAD

**Discussion by alexkunin** [gravitational/teleport#43392] [Q&A] [answered]
Hey guys,

I have a problem I'm hammering my head against for some time now. My goal is to be able to create my own app and db resources that refer to services running inside a Kubernetes cluster, without using NodePort service type and auto-discovery. But I can make it work, and getting DNS-related errors.

This is how it looks in the logs when I do something like `tsh db connect <db-name>`:
```
Jun 23 20:25:20 my-server teleport[136688]: 2024-06-23T20:25:20Z INFO [CA]        Generating TLS certificate <skip>
Jun 23 20:25:20 my-server teleport[136688]: 2024-06-23T20:25:20Z INFO [CA]        Generating TLS certificate <skip>
Jun 23 20:25:20 my-server teleport[136688]: 2024-06-23T20:25:20Z ERRO [DB:SERVIC] Failed to handle connection. addr:<skip>:57721 error:[
Jun 23 20:25:20 my-server teleport[136688]: ERROR REPORT:
Jun 23 20:25:20 my-server teleport[136688]: Original Error: *net.DNSError lookup <skip>.<skip>.svc.cluster.local: Temporary failure in name resolution
Jun 23 20:25:20 my-server teleport[136688]: Stack Trace:
Jun 23 20:25:20 my-server teleport[136688]:         github.com/gravitational/teleport/lib/srv/db/common/errors.go:161 github.com/gravitational/teleport/lib/srv/db/common.ConvertConnectError
Jun 23 20:25:20 my-server teleport[136688]:         <skip>
Jun 23 20:25:20 my-server teleport[136688]: User Message: lookup <skip>.<skip>.svc.cluster.local: Temporary failure in name resolution] db/server.go:959
```
Similar problem occurs when I open an application:
```
Jun 23 20:49:22 my-server teleport[136688]: 2024-06-23T20:49:22Z ERRO [APP:SERVI] Error forwarding to http://<skip>.<skip>.svc.cluster.local/, err: dial tcp: lookup <skip>.<skip>.svc.cluster.local: Temporary failure in name resolution reverseproxy/reverse_proxy.go:218
```
My setup looks like this:

1. Host 1:
  * Teleport installed as a systemd unit directly on the host
  * k0s controller
  * Teleport Agent installed as a Helm chart
2. Host 2 and Host 3:
  * k0s worker

I use Terraform to wind up some services (e.g. Postgres, Redis, code-server) and manually define Teleport resources. At different points in time services were running on k0s controller (which was in single mode at that time) and worker nodes.

If I define these resources with URI based on svc.cluster.local, I see DNS errors mentioned above.

If I allow agent to discover applications, I can see that generated Teleport resources contain svc.cluster.local URIs, and these auto-discovered applications work. I couldn't get DB auto-discovery working, so can't report on that.

I've switched agent image to `-debug` version and was able to confirm that cluster DNS resolution works fine there.

My guess is, auto-discovered apps are in some way "bound" to agent, and thus benefit from in-cluster DNS resolution, while my manually created apps are "bound" to Teleport cluster controller which uses host's (not Kubernetes') DNS resolution and thus fails to do proper look-ups.

Thanks!

**Comment by webvictim** [comment]
Could you please you share some example Teleport configs/values that would allow reproduction of your issue?

I think you are partly correct, but my understanding is that the `app_service` will always resolve DNS locally on the host where it's running. It'd be good to know the exact circumstances where lookups fail so we can investigate further.

**Reply by ak-purplesmart-ai** [reply]
Sure, here is a sample TF snippet:
```terraform
resource "kubernetes_namespace" "test" {
  metadata {
    name = "test"
  }
}

resource "kubernetes_pod" "test" {
  metadata {
    name      = "test"
    namespace = kubernetes_namespace.test.metadata.0.name
    labels = {
      app = "test"
    }
  }
  spec {
    container {
      image = "nginx"
      name  = "test"
    }
  }
}

resource "kubernetes_service" "test" {
  metadata {
    name      = "test"
    namespace = kubernetes_namespace.test.metadata.0.name
  }
  spec {
    selector = {
      app = kubernetes_pod.test.metadata.0.labels.app
    }
    type = "ClusterIP"
    port {
      port = 80
    }
  }
}

resource "teleport_app" "test" {
  version = "v3"

  metadata = {
    name = "test"
  }

  spec = {
    uri = "http://test.${kubernetes_service.test.metadata.0.name}.svc.cluster.local"
  }
}
```
And here are the logs:
```
Jun 24 16:35:35 <skip> teleport[136688]: 2024-06-24T16:35:35Z ERRO [APP:SERVI] Error forwarding to http://test.test.svc.cluster.local/, err: dial tcp: lookup test.test.svc.cluster.local: Temporary failure in name resolution reverseproxy/reverse_proxy.go:218
Jun 24 16:35:35 <skip> teleport[136688]: 2024-06-24T16:35:35Z INFO [APP:WEB]   Round trip: GET /, code: 500, duration: 3.185979ms tls:version: 304, tls:resume:true, tls:csuite:1301, tls:server:<skip>  reverseproxy/reverse_proxy.go:223
```
Nothing special about Teleport or Kubernetes provider, I'm using what official [docs](https://goteleport.com/docs/management/dynamic-resources/terraform-provider/) suggest.

Running `tsh proxy app test` unsurprisingly gives the same result (both browser page and logs).

Running `kubectl -n test port-forward svc/test 8000:80` gives proper "Welcome to nginx!" page in the browser, so, the pod and service work fine and are reachable by means other than Teleport.

**Reply by ak-purplesmart-ai** [reply]
BTW 136688 is the PID of master Teleport instance which runs directly on host (and I'm looking at the logs using `journalctl`).

Logs of Teleport agent which is running inside Kubernetes show this (and this was only once, probably when I've clicked "Launch" in Teleport UI, but not really sure about that). Attempts to open a proxy or reload proxied page do not cause any additional agent logs.

**Reply by webvictim** [reply]
Hey - thanks, but I more meant in terms of the `/etc/teleport.yaml` config file for your base cluster, and also the values for the app agent (`app_service`) that you have running.

It looks like the apps are essentially being dynamically registered (https://goteleport.com/docs/application-access/guides/dynamic-registration) but the `app_service` that's trying to serve the apps isn't running inside the Kubernetes cluster so the DNS resolution is failing. Some more insight into the values and config will help track this down.

**Reply by ak-purplesmart-ai** [reply]
And this is Teleport agent's configuration as I see it from Kubernentes Dashboard:
```yaml
app_service:
  enabled: true
  resources:
  - labels:
      teleport.dev/kubernetes-cluster: cluster-1
      teleport.dev/origin: discovery-kubernetes
auth_service:
  enabled: false
db_service:
  enabled: false
discovery_service:
  discovery_group: cluster-1
  enabled: true
  kubernetes:
  - labels:
      <skip>/exposed-app: \"true\"
    namespaces:
    - '*'
    types:
    - app
kubernetes_service:
  enabled: true
  kube_cluster_name: cluster-1
  labels:
    teleport.internal/resource-id: <skip>
proxy_service:
  enabled: false
ssh_service:
  enabled: false
teleport:
  join_params:
    method: token
    token_name: /etc/teleport-secrets/auth-token
  log:
    format:
      extra_fields:
      - timestamp
      - level
      - component
      - caller
      output: text
    output: stderr
    severity: INFO
  proxy_server: <skip>:443
version: v3
```

**Reply by ak-purplesmart-ai** [reply]
Well, I might have answered my own question. After I've added labels from agent config proxy seems to be working fine:
```terraform
resource "teleport_app" "test" {
  version = "v3"

  metadata = {
    name = "test"
    labels = {
      "teleport.dev/kubernetes-cluster" = "cluster-1"
      "teleport.dev/origin" = "discovery-kubernetes"
    }
  }

  spec = {
    uri = "http://test.${kubernetes_service.test.metadata.0.name}.svc.cluster.local"
  }
}
```
Need to test if it's reliably working as expected.

Another finding, `teleport.dev/origin: discovery-kubernetes` label behaves similarly to `teleport.dev/origin: dynamic`: Teleport automatically adds it, but in .tf have to add it too otherwise Terraform gets confused by this label being added by third-party.

**Reply by webvictim** [reply]
Yes - as you've noticed, the `app_service` running inside the cluster will only handle traffic for apps with the `teleport.dev/kubernetes-cluster: cluster-1` and `teleport.dev/origin: discovery-kubernetes` labels.

**Reply by ak-purplesmart-ai** [reply]
I guess this is the answer I was looking for. Thank you for your help!

**Reply by ak-purplesmart-ai** [reply]
And, if somebody looks for the answer to similar question regarding DBs, the principle is the same -- add labels mentioned above to `teleport_database` resources, but also:
1. You need to enable agent's `db_service` by adding `db` role to agent's helm chart values along with this snippet:
```yaml
databaseResources:
  - labels:
      '*': '*'
```
2. You need to generate a join token which for all roles mentioned in helm char values, e.g. standard Kubernetes cluster procedure does not include `db` role.

**Reply by ak-purplesmart-ai** [reply]
Aaand... it did not help. The issue comes and goes randomly, I have two service/pod pairs run next to each other (created by same module with different configuration which affects only names or resources), one of them is working fine, the other one causes DNS resolution errors in logs.

As far as I can tell, one is served by Teleport agent from inside of Kubernetes, the other one is served by main Teleport instlalation which is outside of Kubernetes.

@webvictim, can you unmark the selected answer, since it does not solve the issue, please?

**Reply by zmb3** [reply, marked as answer]
@webvictim's answer is correct.

You appear to be using dynamic registration. I recommend having a read through [the docs](https://goteleport.com/docs/application-access/guides/dynamic-registration/) to better understand how this feature works.

In a nutshell, you have registered an app named `test` with terraform that contains the following labels:

```
  metadata = {
    name = "test"
    labels = {
      "teleport.dev/kubernetes-cluster" = "cluster-1"
      "teleport.dev/origin" = "discovery-kubernetes"
    }
  }
```

Your app service is declaring that it can proxy the traffic _for any and all apps_ with these labels:

```
app_service:
  enabled: true
  resources:
  - labels:
      teleport.dev/kubernetes-cluster: cluster-1
      teleport.dev/origin: discovery-kubernetes
```

If you are running multiple `app_service`s with the same configuration (and it sounds like you are), then they will each handle traffic for the `test` app. If some of these app services are not able to resolve DNS for the app (because they're running outside the kube cluster where the kube service exists), then you would get exactly the behavior you describe.

To fix this, you need to make sure that only the `app_services` that can actually handle apps in kube cluster `cluster-1` are declaring this label in their configuration.

**Reply by ak-purplesmart-ai** [reply]
Thank you, @zmb3, now it's clear what is wrong. The missing bit was: if several app services match app via labels, any of them can get the ownership of said app. For some reason (ahem, who said CSS...) I was sure that more specific selector wins.

Also, it seems application resources linger a bit in the system after deletion, so it wasn't always clear if I'm working with the latest configuration. Noticed this many times after doing `tctl rm` and then seeing some apps in Teleport UI and `tsh app ls` output.

Anyhow, I hope this explanation and solution will finally work, and won't need to get back to this discussion.

Thanks again!

P.S. I've seen that doc page during my installation and troubleshooting sessions, it seems all dots are there -- but not connections, and I couldn't connect the dots on my own.
